Import/export using myjson.com

elxeno opened the pull request:
ok, first time trying to do this, let me know if something is wrong
- changed createJSON function to return the string so it can be used on export function
- created functions for importing (from address like GET) and exporting (new button on menu)

issues:
- need to fix css menu to fit extra button (on small screen) 
- create a text display for the new address (don't know how to do it in the new format)
- ~~a few problems importing data when width != height (tried with boilers, probably related to issue #22)~~ looks like #29 fixed it

elxeno committed:
created export/import functions

using myjson.com
need to fix css for extra button and create a display for the new address

elxeno committed:
removed logging and added some comments

elxeno committed:
fixed rotation on load

lessthantrue commented:
Is there a reason why we are sending the generated blueprint to myJSON instead of storing it locally?
Also I'm wondering if we need to store the blueprints in JSON form at all. Does the blueprint library have a reverse encoder, or something that takes the blueprint string and turns it back into JSON?

elxeno commented:
trying to do it in a way that it doesnt need a server, so it could continue to be hosted on github or run on pc (with file://), is it a bad idea?

elxeno commented:
also i think this way we dont need to worry about security, is that right?

lessthantrue commented:
You're right that it simplifies hosting when we avoid having assets that are dynamically saved, loaded, and deleted. I guess the question is what is the "save blueprint" button supposed to do, and if possible, can it be done without essentially hotlinking myJSON?

elxeno commented:
I'm happy with the result, although the code is inefficient, my idea was for it to work kinda like imgur/gfycat etc.. U click save and get an url to share, and when someone goes to that url it loads the blueprint.. It worked well on my tests, but right now it only sends an alert with the url, i wanted to create a window like the one for the blueprint string, but i don't understand the changes that were made on the layout so i rather not mess with that..

Edit: about the myjson thing, at first i wanted to use pastebin or something similar, but that would have to expose the API key(using js only), and myjson is the only one i found that doesn't require a key... Any ideas about this issue?

lessthantrue commented:
I see. It didn't make too much sense to me when all I got was the alert, but it's a good idea. I was thinking more along the lines of saving a blueprint for use later by the same user.

Here's my suggestion for what might at this point be a separate enhancement: we already have users sharing blueprints via blueprint strings. Other projects have ways of turning these strings back into JSON (see: https://github.com/trakos/fbpviewer , MIT license), which we could easily also implement. What we would do is have the user enter this blueprint string into the text box, which the script would read, decode, and import into the editing field.

elxeno commented:
Yeah that could be added easily, just need another textbox, The code i used for the encoder also had a decoder, and it uses the same libraries that are already included, i'll include it later, just need to decode and use the same function to load everything.. 
But theres also a few issues with this.. If blueprint was not created on the site, the size of blueprint and position of the entities might be outside the tiles, and when blueprints from game include rails they give integer positions to rails and offsets everything else by 0.5 (it's usually *.5 for every item that has width or height of 2, 4, etc). I left the rails with decimal numbers and it works the same (probably because they will snap to grid when in game)

Commit:
added modal to blueprint save button
this replaces the old alert
Note: the textarea for this modal does not automatically scale, I don't
know how to fix that.

lessthantrue commented:
For now, the blueprint size issue could have as simple a fix as an alert that says "blueprint too large" until we get scaling the editing field figured out.

Zeragamba commented:
Adding my (late) thoughts: I don't think we want to tie this project into some external storage site that may not exist later. If we want persistent storage, Local Storage in the browser should be good enough.

elxeno commented:
Should be fine to keep it while we work on other solutions right? If it helps it is on github too https://github.com/lramoth/myjson under MIT license

elxeno committed:
Merge pull request #31 from lessthantrue/import-export

added modal to blueprint save button

Zeragamba commented on js/script.js:
```diff
@@ -11,6 +11,12 @@ window.onload = function () {
     createItems();
     createTiles();
+    // https://stackoverflow.com/questions/1586330/access-get-directly-from-javascript#1586333
+    window.$_GET = location.search.substr(1).split("&").reduce((o,i)=>(u=decodeURIComponent,[k,v]=i.split("="),o[u(k)]=v&&u(v),o),{});
```
This line uses some ES6 features, which are not supported in IE11: http://kangax.github.io/compat-table/es6/

do we want to keep support for IE11?
 
Also, this line is very very hard to understand, can we please split this up into something readable?

elxeno replied:
separated the code and replaced arrow function and return, did that fix it?
